test/Test/Cardano/Prelude/GHC/Heap/Size.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MagicHash                  #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE TemplateHaskell            #-}
+
+{-# OPTIONS_GHC -fno-full-laziness #-}
+
+module Test.Cardano.Prelude.GHC.Heap.Size (tests) where
+
+import Cardano.Prelude hiding (diff)
+
+import GHC.Prim (Int#)
+import qualified Data.ByteString as BS.Strict
+
+import Hedgehog
+
+{-------------------------------------------------------------------------------
+  Various unboxed tuples, for testing purposes
+
+  We don't use an UNPACK pragma here because that is ignored at @-O0@, which
+  could result in false negatives.
+-------------------------------------------------------------------------------}
+
+data Ints2 = Ints2 Int# Int#
+data Ints3 = Ints3 Int# Int# Int#
+data Ints4 = Ints4 Int# Int# Int# Int#
+data Ints5 = Ints5 Int# Int# Int# Int# Int#
+
+{-------------------------------------------------------------------------------
+  Auxiliary
+-------------------------------------------------------------------------------}
+
+newtype NumWords = NumWords Int
+  deriving (Show, Eq, Ord, Num)
+
+-- | Size of a word (in bytes)
+wordSize :: Int
+wordSize = finiteBitSize (0 :: Word) `div` 8
+
+-- | Convert number of bytes to number of machine words
+bytesToNumWords :: Word64 -> NumWords
+bytesToNumWords n = NumWords (fromIntegral n `divRoundUp` wordSize)
+
+-- | Divide @n@ by @m@, rounding up if necessary
+--
+-- > divRoundUp 10 2 == 5
+-- > divRoundUp 10 3 == 4
+divRoundUp :: Int -> Int -> Int
+divRoundUp n m = ceiling (n' / m')
+  where
+    n', m' :: Double
+    n' = fromIntegral n
+    m' = fromIntegral m
+
+-- | Estimate the size of a strict bytestring
+--
+-- Strict 'ByteString' is defined as
+--
+-- > data ByteString = PS {-# UNPACK #-} !(ForeignPtr Word8) -- payload
+-- >                      {-# UNPACK #-} !Int                -- offset
+-- >                      {-# UNPACK #-} !Int                -- length
+--
+-- where
+--
+-- > data ForeignPtr a = ForeignPtr Addr# ForeignPtrContents
+-- > data ForeignPtrContents
+-- >   = PlainForeignPtr !(IORef Finalizers)
+-- >   | MallocPtr      (MutableByteArray# RealWorld) !(IORef Finalizers)
+-- >   | PlainPtr       (MutableByteArray# RealWorld)
+--
+-- The UNPACK pragma on the 'ForeignPtr' means that the @Addr#@ from
+-- 'ForeignPtr' will be inlined in the 'PS' constructor, resulting in a 5-word
+-- size for 'ByteString' and a 2-word size for 'ForeignPtr' (including the info
+-- table pointer). For regularly constructed bytestring only uses 'PlainPtr'.
+--
+-- Since the payload itself will be allocated rounded to the nearest word, a
+-- bytestring of @n@ elements is @5 + 2 + 2 + ceiling (n / wordSize)@ words.
+bsSize :: Int -> NumWords
+bsSize numElems = NumWords (5 + 2 + 2 + numElems `divRoundUp` wordSize)
```

Comment:
Worth noting that we already have a `HeapWords` instance for estimating the size of a `ByteString`:

https://github.com/input-output-hk/cardano-prelude/blob/4c48d34fd60991bb1b9168d293cfe31a975d858e/src/Cardano/Prelude/HeapWords.hs#L513-L515

Reply:
That is incorrect :)